### DISCUSSION TITLE
Support for nested template in Datasource

### DISCUSSION THREAD

**Discussion by HenryXie1** [hairyhenderson/gomplate#1239] [Q&A] [answered]
We are using vault as datasource on gomplate
The secret path we would like to fetch from config yaml file.
There is nested template case here 

It is working with hardcode
```
{{ (ds "vault:" "versioned/data/testdata").data.client_id }}
```
config yaml is 
```
spec:
   vault:
     secretPrefix: versioned/data/testdata
```
Not sure how we can contruct `secretPrefix` from ds, below is not working
` {{ (ds "vault:" {{ (ds "config").spec.vault.secretPrefix }}).data.client_id }}`

wonder if it is supported or any workaround? Thanks heaps

**Comment by HenryXie1** [comment]
Add a variable but it does not work either
 ` {{ $prefix := "{{ (ds \"config\").spec.vault.secretPrefix }}" }} {{ (ds "vault:" $prefix).data.client_id }}`

**Comment by hairyhenderson** [comment, marked as answer]
Hi @HenryXie1, thanks for filing this - I've converted this to a discussion as this is more of a usage question than an issue. If there is indeed a bug or feature request that comes out of this, we can open an issue from it later!

That said, the syntax `{{ (ds "vault:" {{ (ds "config").spec.vault.secretPrefix }}).data.client_id }}` obviously doesn't work - this is because everything between the outer `{{}}` is in a template context already, and inner `{{}}` won't change that.

You can just reference the `ds` function instead:

```handlebars
{{ (ds "vault:" (ds "config").spec.vault.secretPrefix).data.client_id }}
```

Or, with a variable (which may be more readable anyway):

```handlebars
{{ $prefix := (ds "config").spec.vault.secretPrefix }}
{{ (ds "vault:" $prefix).data.client_id }}
```

To make things slightly more readable, and assuming the `config` datasource is static (unlike the dynamic inline vault datasource you're creating), you can use the context instead:

```handlebars
{{ $prefix := .config.spec.vault.secretPrefix }}
{{ (ds "vault:" $prefix).data.client_id }}
```

you'd call `gomplate` with `-c` to read the `config` datasource into the context:

```console
$ gomplate -c config=path/to/config.yaml -f mytemplate.tmpl
```

 (see [the docs](https://docs.gomplate.ca/usage/#context-c) for more details)

Hope this helps - please mark this as answered if it solves your issue.
